nemo/core/utils/k2_guard.py

```diff
@@ -26,39 +38,20 @@
     "It is advised to always install k2 using setup.sh only, "
     "as different versions of k2 may not interact with the NeMo code as expected."
 )
-K2_IMPORT_SUCCESS = "Successfully imported k2."
-
 
-def k2_is_available() -> Tuple[bool, str]:
-    try:
-        import k2
+if not package_available("k2"):
+    raise ModuleNotFoundError(K2_INSTALLATION_MESSAGE)
```

Comment:
How is it an optional dependency if merely importing it will raise an error ? K2 is not installed by default, the check should be don't at init inside of where it will be used rather than during simply import of a file.

Reply:
Currently, k2 is not included in the import path of init.py files for this reason, but at least the check is done at init of the class. This instead puts it in the import path so if used simply imports the file it will crash with helpful error. 

Numba checks specifically defers this crash to the moment user actually tries to use the module rather than simply import the module

Reply:
Importing `k2` raises an error only when importing `k2_guard` of `k2 from k2_guard`
`k2_guard` will not be imported and cached.

So, every module that tries to import `k2 from k2_guard` will raise an error (if no k2 / incorrect version) and will not be initialized (no classes in this module).
As a result, importing from **any** of the modules in nemo, that depend on `k2` (maybe not directly) will result in error with the necessary message, which is a correct behavior.

But importing not at the top level will allow to raise an error in runtime, we do not need to call `k2_import_guard()`:

```python
if backend == "k2":
    if criterion_type == "ml":
        from nemo.collections.asr.parts.k2.ml_loss import MLLoss as K2Loss  # maybe error here
    elif criterion_type == "map":
        from nemo.collections.asr.parts.k2.map_loss import MAPLoss as K2Loss  # or here
```

Also, **this doesn't change the previous behavior**: in modules `k2_import_guard()` call was used at the top level, so incorrect k2 version resulted in immediate crash **before** importing any defined class.

Reply:
You can set `__K2_MINIMUM_MINOR_VERSION = 111` and see that the behavior didn't change and everything works.